lib/SILGen/SILGenApply.cpp

```diff
@@ -627,6 +628,16 @@ class Callee {
       return fn;
     }
     case Kind::WitnessMethod: {
+      if (auto func = constant->getFuncDecl()) {
+        if (func->isDistributed() && isa<ProtocolDecl>(func->getDeclContext())) {
+          // If we're calling cross-actor, we must always use a distributed thunk
+          if (!isSameActorIsolated(func, SGF.FunctionDC)) {
+            // We must adjust the constant to use a distributed thunk.
+            constant = constant->asDistributed();
```

Comment:
This doesn't seem like the right place to select the distributed thunk. It should be driven by `ApplyExpr::shouldApplyDistributedThunk()`, when SILGenApply decomposes a call.

Reply:
Also, I expect that the `isSameActorIsolated` check is wrong in cases where you're calling a distributed requirement from (say) an inheriting protocol:

```swift
protocol P: DistributedActor {
  distributed func f()
}

protocol Q: P { }

extension Q {
  func g() {
    f()
  }
} 
```

You would expect `isSameActorIsolated` to be true, but it isn't because the `actor` decls are different. The type checker understands this and knows that this isn't a call through a distributed thunk.

Reply:
Oh I see, thanks for the example case -- I'll dig into this in a follow up PR 👍